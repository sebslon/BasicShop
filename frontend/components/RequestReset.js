import gql from "graphql-tag";
import { useMutation } from "@apollo/client";

import Form from "./styles/Form";
import useForm from "../lib/useForm";
import DisplayError from "./ErrorMessage";

export const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      code
      message
    }
  }
`;

export default function RequestReset() {
  const { inputs, handleChange, resetForm } = useForm({
    email: "",
  });

  const [requestReset, { data, loading, error }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: inputs,
    }
  );

  async function handleSubmit(e) {
    e.preventDefault();

    await requestReset().catch(console.error);
    resetForm();
  }

  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Request a Password Reset</h2>
      <DisplayError error={error} />
      <fieldset>
        {data?.sendUserPasswordResetLink === null && (
          <p>Check your email for reset password link!</p>
        )}
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="Your Email Address"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>

        <button type="submit">Submit</button>
      </fieldset>
    </Form>
  );
}
