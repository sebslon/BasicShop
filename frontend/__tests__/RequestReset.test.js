import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import userEvent from "@testing-library/user-event";

import RequestReset, {
  REQUEST_RESET_MUTATION,
} from "../components/RequestReset";

const email = "test@gmail.com";
const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email },
    },
    result: {
      data: { sendUserPasswordResetLink: null },
    },
  },
];

describe("<RequestReset />", () => {
  it("Renders and matches snapshot", () => {
    const { container } = render(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("Calls the mutation when submitted", async () => {
    const { container } = render(
      <MockedProvider>
        <RequestReset mocks={mocks} />
      </MockedProvider>
    );

    userEvent.type(screen.getByPlaceholderText(/email/i), email);
    userEvent.click(screen.getByText(/submit/i));

    const success = await screen.findByText(/success/i);

    expect(success).toBeInTheDocument();
  });
});
