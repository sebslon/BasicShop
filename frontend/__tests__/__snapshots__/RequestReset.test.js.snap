// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset /> Renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1wuhrih-0 fUbzWs"
    method="POST"
  >
    <h2>
      Request a Password Reset
    </h2>
    <fieldset>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Your Email Address"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Submit
      </button>
    </fieldset>
  </form>
</div>
`;
